rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,142 @@
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+   * The pairs must be on the same line (unless a `\` escapes the newline) but expressions needn't be on the same line.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting specification may follow the expression, delimited by an unambiguous character.
+   * Restriction: the formatting specification must be constant at parse time.
+   * The syntax for formatting specification and a delimiter is undefined in this RFC. A future RFC will implement this.
+
+To put the above into formal EBNF grammar:
+
+```
+stringinterp ::= <INTERP_BEGIN> exp {<INTERP_MID> exp} <INTERP_END>
+```
+
+Which, in actual Luau code, will look like the following:
+
+```
+local world = "world"
+print(`Hello {world}!`)
+--> Hello world!
+
+local combo = {5, 2, 8, 9}
+print(`The lock combinations are: {table.concat(combo, ", ")}`)
+--> The lock combinations are: 5, 2, 8, 9
+
+local set1 = Set.new({0, 1, 3})
+local set2 = Set.new({0, 5, 4})
+print(`{set1} ∪ {set2} = {Set.union(set1, set2)}`)
+--> {0, 1, 3} ∪ {0, 5, 4} = {0, 1, 3, 4, 5}
+
+print(`Some example escaping the braces %{like so}`)
+print(`% that doesn't escape anything is part of the string...`)
+print(`%% will escape the second %...`)
+print(`Some text that also includes \`...`)
+--> Some example escaping the braces {like so}
+--> % that doesn't escape anything is part of the string...
+--> % will escape the second %...
+--> Some text that also includes `...
+```
+
+As for how newlines are handled, they are handled the same as other string literals. Any text between the `{}` delimiters are not considered part of the string, hence newlines are OK. The main thing is that one opening pair will scan until either a closing pair is encountered, or an unescaped newline.
+
+```
+local name = "Luau"
+
+print(`Welcome to {
+    Luau
+}!`)
+--> Welcome to Luau!
+
+print(`Welcome to \
+{Luau}!`)
+--> Welcome to
+--  Luau!
+```
+
+This expression will not be allowed to come after a `prefixexp`. I believe this is fully additive, so a future RFC may allow allow this. So for now, we explicitly reject the following:
+
+```
+local name = "world"
+print`Hello {world}`
+```
+
+Since the string interpolation expression is going to be lowered into a `string.format` call, we'll also need to extend `string.format`. The bare minimum to support the lowering is to add a new token whose definition is to perform a `tostring` call. `%a` is currently an invalid token, so this is a backward compatible extension. `%a` will have the same behavior as if `tostring` was called.
```

Comment:
`%a` and `%A` are valid in Lua 5.3+, and don't do what you described.
```lua
print(string.format("%a",1.1))
-- in Lua 5.3+ this prints: 0x1.199999999999ap+0
-- in Luau this would print: 1.1
```
Something else should be used to allow for the Lua 5.3+ behavior to be backported and to avoid a weird difference between Lua 5.3+ and Luau.

Reply:
Changed it to `%*`. I had thought that hexadecimal floats were something we didn't plan to do. I was mistaken. `%*` it is. We're not likely to support that one.